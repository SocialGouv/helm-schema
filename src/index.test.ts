import { expect, test } from "@jest/globals";

import { extractValues, toJsonSchema } from "./index";

const tests = [
  {
    title: "YAML without comment",
    yaml: `
name: Jul
age: 42

location:
  country: FR
  `,
  },
  {
    title: "YAML with comments",
    yaml: `
# @param {string} name Your name
name: Jul
# @param {number} age Your age
age: 42

location:
  # @param {string} [country] Your country
  country: FR
  `,
  },
  {
    title: "YAML with nestedcomments",
    yaml: `
family:
  # @param {object} mother The mother
  mother:
    # @param {string} name The mother's name
    name:
    # @param {object} [mother] The mother's mother
    mother:
      # @param {string} [name] The mother mother's name
      name:
`,
  },
  {
    title: "YAML with sections",
    yaml: `
# @section family blablab
family:
  # @param {object} [mother] The mother
  mother:
  `,
  },
  {
    title: "single-line",
    yaml: `
# @param {number} number The magic number
number: 42
  `,
  },
  {
    title: "YAML with multiline comment",
    yaml: `
# family blablab
# more info about that crazy family
family:
  # the mother is important in your life
  # @param {object} [mother] The mother
  mother:
  `,
  },
];

tests.forEach((t) => {
  test(`extractValues: ${t.title}`, () => {
    expect(extractValues(t.yaml)).toMatchSnapshot();
  });
});

tests.forEach((t) => {
  test(`toJsonSchema: ${t.title}`, () => {
    expect(toJsonSchema(t.yaml)).toMatchSnapshot();
  });
});

test("toJsonSchema: add root properties", () => {
  expect(
    toJsonSchema(
      `
# @param {string} [some] Some optional string
some: thing`,
      {
        $id: "some-id",
        title: "schema title",
      }
    )
  ).toMatchSnapshot();
});
