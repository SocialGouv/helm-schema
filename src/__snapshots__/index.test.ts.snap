// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractValues: YAML with comments 1`] = `
[
  {
    "comment": {
      "description": "Your name",
      "key": "name",
      "required": true,
      "type": "string",
    },
    "key": "name",
    "value": "Jul",
  },
  {
    "key": "age",
    "value": "42",
  },
  {
    "children": [
      {
        "comment": {
          "description": "Your country",
          "key": "country",
          "required": false,
          "type": "string",
        },
        "key": "country",
        "value": "FR",
      },
    ],
    "key": "location",
    "type": "object",
  },
]
`;

exports[`extractValues: YAML with nestedcomments 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "comment": {
              "description": "The mother's name",
              "key": "name",
              "required": true,
              "type": "string",
            },
            "key": "name",
          },
          {
            "children": [
              {
                "comment": {
                  "description": "The mother mother's name",
                  "key": "name",
                  "required": false,
                  "type": "string",
                },
                "key": "name",
              },
            ],
            "comment": {
              "description": "The mother's mother",
              "key": "mother",
              "required": false,
              "type": "object",
            },
            "key": "mother",
            "type": "object",
          },
        ],
        "comment": {
          "description": "The mother",
          "key": "mother",
          "required": true,
          "type": "object",
        },
        "key": "mother",
        "type": "object",
      },
    ],
    "key": "family",
    "type": "object",
  },
]
`;

exports[`extractValues: YAML with sections 1`] = `
[
  {
    "children": [
      {
        "comment": {
          "description": "The mother",
          "key": "mother",
          "required": false,
          "type": "object",
        },
        "key": "mother",
      },
    ],
    "comment": {
      "description": "family blablab",
    },
    "key": "family",
    "type": "object",
  },
]
`;

exports[`extractValues: YAML without comment 1`] = `
[
  {
    "key": "name",
    "value": "Jul",
  },
  {
    "key": "age",
    "value": "42",
  },
  {
    "children": [
      {
        "key": "country",
        "value": "FR",
      },
    ],
    "key": "location",
    "type": "object",
  },
]
`;

exports[`extractValues: single-line 1`] = `
[
  {
    "comment": {
      "description": "The magic number",
      "key": "number",
      "required": true,
      "type": "number",
    },
    "key": "number",
    "value": "42",
  },
]
`;

exports[`toJsonSchema: YAML with comments 1`] = `
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "properties": {
    "age": {
      "default": "42",
      "type": "string",
    },
    "location": {
      "properties": {
        "country": {
          "default": "FR",
          "description": "Your country",
          "type": "string",
        },
      },
      "required": [],
      "type": "object",
    },
    "name": {
      "default": "Jul",
      "description": "Your name",
      "type": "string",
    },
  },
  "required": [
    "name",
  ],
  "type": "object",
}
`;

exports[`toJsonSchema: YAML with nestedcomments 1`] = `
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "properties": {
    "family": {
      "properties": {
        "mother": {
          "description": "The mother",
          "properties": {
            "mother": {
              "description": "The mother's mother",
              "properties": {
                "name": {
                  "description": "The mother mother's name",
                  "type": "string",
                },
              },
              "required": [],
              "type": "object",
            },
            "name": {
              "description": "The mother's name",
              "type": "string",
            },
          },
          "required": [
            "name",
          ],
          "type": "object",
        },
      },
      "required": [
        "mother",
      ],
      "type": "object",
    },
  },
  "required": [],
  "type": "object",
}
`;

exports[`toJsonSchema: YAML with sections 1`] = `
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "properties": {
    "family": {
      "description": "family blablab",
      "properties": {
        "mother": {
          "description": "The mother",
          "type": "object",
        },
      },
      "required": [],
      "type": "object",
    },
  },
  "required": [],
  "type": "object",
}
`;

exports[`toJsonSchema: YAML without comment 1`] = `
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "properties": {
    "age": {
      "default": "42",
      "type": "string",
    },
    "location": {
      "properties": {
        "country": {
          "default": "FR",
          "type": "string",
        },
      },
      "required": [],
      "type": "object",
    },
    "name": {
      "default": "Jul",
      "type": "string",
    },
  },
  "required": [],
  "type": "object",
}
`;

exports[`toJsonSchema: add root properties 1`] = `
{
  "$id": "some-id",
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "properties": {
    "some": {
      "default": "thing",
      "description": "Some optional string",
      "type": "string",
    },
  },
  "required": [],
  "title": "schema title",
  "type": "object",
}
`;

exports[`toJsonSchema: single-line 1`] = `
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "properties": {
    "number": {
      "default": "42",
      "description": "The magic number",
      "type": "number",
    },
  },
  "required": [
    "number",
  ],
  "type": "object",
}
`;
